// @ceramicnetwork/http-client/lib/remote-admin-api.js
import { fetchJson, } from '@ceramicnetwork/common';
import { RemotePinApi } from './remote-pin-api.js';
import { StreamID } from '@ceramicnetwork/streamid';
import { MissingDIDError } from './utils.js';
export class RemoteAdminApi {
    constructor(_apiUrl, _getDidFn) {
        this._apiUrl = _apiUrl;
        this._getDidFn = _getDidFn;
        this._fetchJson = fetchJson;
        this.modelsPath = './admin/models';
        this.modelDataPath = './admin/modelData';
        this.getCodePath = './admin/getCode';
        this.nodeStatusPath = './admin/status';
        this._pinApi = new RemotePinApi(this._apiUrl, this._getDidFn);
    }
    getCodeUrl() {
        return new URL(this.getCodePath, this._apiUrl);
    }
    getModelsUrl() {
        return new URL(this.modelsPath, this._apiUrl);
    }
    getModelDataUrl() {
        return new URL(this.modelDataPath, this._apiUrl);
    }
    getStatusUrl() {
        return new URL(this.nodeStatusPath, this._apiUrl);
    }
    async buildJWS(actingDid, code, requestPath, body) {
        if (!actingDid)
            throw new MissingDIDError();
        const jws = await actingDid.createJWS({
            code: code,
            requestPath,
            requestBody: body,
        });
        return `${jws.signatures[0].protected}.${jws.payload}.${jws.signatures[0].signature}`;
    }
    async generateCode() {
        return (await this._fetchJson(this.getCodeUrl())).code;
    }
    async nodeStatus() {
        const code = await this.generateCode();
        return this._fetchJson(this.getStatusUrl(), {
            headers: {
                Authorization: `Basic ${await this.buildJWS(this._getDidFn(), code, this.getStatusUrl().pathname)}`,
            },
        });
    }
    async startIndexingModels(modelsIDs) {
        const code = await this.generateCode();
        const body = {
            models: modelsIDs.map((idx) => idx.toString()),
        };
        await this._fetchJson(this.getModelsUrl(), {
            method: 'POST',
            body: {
                jws: await this.buildJWS(this._getDidFn(), code, this.getModelsUrl().pathname, body),
            },
        });
    }
    async startIndexingModelData(modelData) {
        const code = await this.generateCode();
        const body = {
            modelData: modelData.map((idx) => {
                return {
                    streamID: idx.streamID.toString(),
                    indices: idx.indices,
                };
            }),
        };
        await this._fetchJson(this.getModelDataUrl(), {
            method: 'POST',
            body: {
                jws: await this.buildJWS(this._getDidFn(), code, this.getModelDataUrl().pathname, body),
            },
        });
    }
    async getIndexedModels() {
        const code = await this.generateCode();
        const response = await this._fetchJson(this.getModelsUrl(), {
            headers: {
                Authorization: `Basic ${await this.buildJWS(this._getDidFn(), code, this.getModelsUrl().pathname)}`,
            },
        });
        return response.models.map((data) => {
            return {
                streamID: StreamID.fromString(data),
            };
        });
    }
    async getIndexedModelData() {
        const code = await this.generateCode();
        const response = await this._fetchJson(this.getModelDataUrl(), {
            headers: {
                Authorization: `Basic ${await this.buildJWS(this._getDidFn(), code, this.getModelDataUrl().pathname)}`,
            },
        });
        return response.modelData.map((data) => {
            return {
                streamID: StreamID.fromString(data.streamID),
                indices: data.indices,
            };
        });
    }
    async stopIndexingModels(modelsIDs) {
        const code = await this.generateCode();
        const body = {
            models: modelsIDs.map((data) => data.toString()),
        };
        await this._fetchJson(this.getModelsUrl(), {
            method: 'DELETE',
            body: {
                jws: await this.buildJWS(this._getDidFn(), code, this.getModelsUrl().pathname, body),
            },
        });
    }
    async stopIndexingModelData(modelData) {
        const code = await this.generateCode();
        const body = {
            modelData: modelData.map((data) => {
                return {
                    streamID: data.streamID.toString(),
                };
            }),
        };
        await this._fetchJson(this.getModelDataUrl(), {
            method: 'DELETE',
            body: {
                jws: await this.buildJWS(this._getDidFn(), code, this.getModelDataUrl().pathname, body),
            },
        });
    }
    get pin() {
        return this._pinApi;
    }
}
