// @smontero/eosio-signing-tools/dist/ChainNodeMap.js
"use strict";

module.exports = {
  '1eaa0824707c8c16bd25145493bf062a': {
    name: 'Telos Testnet',
    host: 'testnet.telos.caleos.io',
    port: 443
  },
  '4667b205c6838ef70ff7988f6e8257e8': {
    name: 'Telos Mainnet',
    host: 'telos.caleos.io',
    port: 443
  },
  aca376f206b8fc25a6ed44dbdc66547c: {
    name: 'EOS Mainnet',
    host: 'api.eosn.io',
    port: 443
  },
  '1064487b3cd1a897ce03ae5b6a865651': {
    name: 'WAX Mainnet',
    host: 'wax.eosn.io',
    port: 443
  },
  d5a3d18fbb3c084e3b1f3fa98c21014b: {
    name: 'BOS Mainnet',
    host: 'bos.eosn.io',
    port: 443
  },
  cfe6486a83bad4962f232d48003b1824: {
    name: 'MEETONE Mainnet',
    host: 'meetone.eosn.io',
    port: 443
  },
  '2a02a0053e5a8cf73a56ba0fda11e4d9': {
    name: 'Jungle Testnet',
    host: 'jungle.eosn.io',
    port: 443
  },
  '5fff1dae8dc8e2fc4d5b23b2c7665c97': {
    name: 'Kylin Testnet',
    host: 'kylin.eosn.io',
    port: 443
  },
  f16b1833c747c43682f4386fca9cbb32: {
    name: 'WAX Testnet',
    host: 'waxtest.eosn.io',
    port: 443
  },
  '33cc2426f1b258ef8c798c34c0360b31': {
    name: 'BOS testnet',
    host: 'bostest.eosn.io',
    port: 443
  }
};
