// @ceramicnetwork/indexing/lib/migrations/cdb-schema-verification.js
import { DatabaseType } from './1-create-model-table.js';
export const STRUCTURES = {
    [DatabaseType.POSTGRES]: {
        RELATION_COLUMN: {
            type: 'character varying',
            maxLength: 1024,
            nullable: false,
            defaultValue: null,
        },
        COMMON_TABLE: {
            stream_id: {
                type: 'character varying',
                maxLength: 255,
                nullable: false,
                defaultValue: null,
            },
            controller_did: {
                type: 'character varying',
                maxLength: 1024,
                nullable: false,
                defaultValue: null,
            },
            stream_content: {
                type: 'jsonb',
                maxLength: null,
                nullable: false,
                defaultValue: null,
            },
            tip: {
                type: 'character varying',
                maxLength: 255,
                nullable: false,
                defaultValue: null,
            },
            last_anchored_at: {
                type: 'timestamp with time zone',
                maxLength: null,
                nullable: true,
                defaultValue: null,
            },
            first_anchored_at: {
                type: 'timestamp with time zone',
                maxLength: null,
                nullable: true,
                defaultValue: null,
            },
            created_at: {
                type: 'timestamp with time zone',
                maxLength: null,
                nullable: false,
                defaultValue: 'CURRENT_TIMESTAMP',
            },
            updated_at: {
                type: 'timestamp with time zone',
                maxLength: null,
                nullable: false,
                defaultValue: 'CURRENT_TIMESTAMP',
            },
        },
        CONFIG_TABLE_MODEL_INDEX: {
            model: {
                type: 'character varying',
                maxLength: 1024,
                nullable: false,
                defaultValue: null,
            },
            is_indexed: {
                type: 'boolean',
                maxLength: null,
                nullable: false,
                defaultValue: 'true',
            },
            enable_historical_sync: {
                type: 'boolean',
                maxLength: null,
                nullable: false,
                defaultValue: 'false',
            },
            created_at: {
                type: 'timestamp with time zone',
                maxLength: null,
                nullable: false,
                defaultValue: 'CURRENT_TIMESTAMP',
            },
            updated_at: {
                type: 'timestamp with time zone',
                maxLength: null,
                nullable: false,
                defaultValue: 'CURRENT_TIMESTAMP',
            },
            updated_by: {
                type: 'character varying',
                maxLength: 1024,
                nullable: false,
                defaultValue: null,
            },
            indices: {
                type: 'jsonb',
                maxLength: null,
                nullable: true,
                defaultValue: null,
            },
        },
        CONFIG_TABLE: {
            option: {
                type: 'character varying',
                maxLength: 1024,
                nullable: false,
                defaultValue: null,
            },
            value: {
                type: 'character varying',
                maxLength: 1024,
                nullable: false,
                defaultValue: null,
            },
            created_at: {
                type: 'timestamp with time zone',
                maxLength: null,
                nullable: false,
                defaultValue: 'CURRENT_TIMESTAMP',
            },
            updated_at: {
                type: 'timestamp with time zone',
                maxLength: null,
                nullable: false,
                defaultValue: 'CURRENT_TIMESTAMP',
            },
            updated_by: {
                type: 'character varying',
                maxLength: 1024,
                nullable: true,
                defaultValue: null,
            },
        },
        MODEL_IMPLEMENTS_TABLE: {
            interface_id: {
                type: 'character varying',
                maxLength: 1024,
                nullable: false,
                defaultValue: null,
            },
            implemented_by_id: {
                type: 'character varying',
                maxLength: 1024,
                nullable: false,
                defaultValue: null,
            },
        },
    },
    [DatabaseType.SQLITE]: {
        RELATION_COLUMN: {
            type: 'varchar',
            maxLength: '1024',
            nullable: false,
            defaultValue: null,
        },
        COMMON_TABLE: {
            stream_id: {
                type: 'varchar',
                maxLength: '1024',
                nullable: false,
                defaultValue: null,
            },
            controller_did: {
                type: 'varchar',
                maxLength: '1024',
                nullable: false,
                defaultValue: null,
            },
            stream_content: {
                type: 'varchar',
                maxLength: '255',
                nullable: false,
                defaultValue: null,
            },
            tip: {
                type: 'varchar',
                maxLength: '255',
                nullable: false,
                defaultValue: null,
            },
            last_anchored_at: {
                type: 'integer',
                maxLength: null,
                nullable: true,
                defaultValue: null,
            },
            first_anchored_at: {
                type: 'integer',
                maxLength: null,
                nullable: true,
                defaultValue: null,
            },
            created_at: {
                type: 'integer',
                maxLength: null,
                nullable: false,
                defaultValue: null,
            },
            updated_at: {
                type: 'integer',
                maxLength: null,
                nullable: false,
                defaultValue: null,
            },
        },
        CONFIG_TABLE_MODEL_INDEX: {
            model: {
                type: 'varchar',
                maxLength: '1024',
                nullable: false,
                defaultValue: null,
            },
            is_indexed: {
                type: 'boolean',
                maxLength: null,
                nullable: false,
                defaultValue: "'1'",
            },
            enable_historical_sync: {
                type: 'boolean',
                maxLength: null,
                nullable: false,
                defaultValue: "'0'",
            },
            created_at: {
                type: 'datetime',
                maxLength: null,
                nullable: false,
                defaultValue: 'CURRENT_TIMESTAMP',
            },
            updated_at: {
                type: 'datetime',
                maxLength: null,
                nullable: false,
                defaultValue: 'CURRENT_TIMESTAMP',
            },
            updated_by: {
                type: 'varchar',
                maxLength: '1024',
                nullable: false,
                defaultValue: null,
            },
            indices: {
                type: 'json',
                maxLength: null,
                nullable: true,
                defaultValue: null,
            },
        },
        CONFIG_TABLE: {
            option: {
                type: 'varchar',
                maxLength: '1024',
                nullable: false,
                defaultValue: null,
            },
            value: {
                type: 'varchar',
                maxLength: '1024',
                nullable: false,
                defaultValue: null,
            },
            created_at: {
                type: 'datetime',
                maxLength: null,
                nullable: false,
                defaultValue: 'CURRENT_TIMESTAMP',
            },
            updated_at: {
                type: 'datetime',
                maxLength: null,
                nullable: false,
                defaultValue: 'CURRENT_TIMESTAMP',
            },
            updated_by: {
                type: 'varchar',
                maxLength: '1024',
                nullable: true,
                defaultValue: null,
            },
        },
        MODEL_IMPLEMENTS_TABLE: {
            interface_id: {
                type: 'varchar',
                maxLength: '1024',
                nullable: false,
                defaultValue: null,
            },
            implemented_by_id: {
                type: 'varchar',
                maxLength: '1024',
                nullable: false,
                defaultValue: null,
            },
        },
    },
};
